File: Brute-Force/One-Pass/2023. Number of Pairs of Strings With Concatenation Equal to Target/numOfPairs.py
# hash map
class Solution:
    def numOfPairs(self, nums: List[str], target: str) -> int:
        output = 0
        dict = collections.Counter(nums)
        # print(dict)
        
        for i in range(len(target)):
            prefix = target[0:i]
            suffix = target[i:len(target)]
            
            if prefix == suffix:
                output += dict[prefix] * (dict[suffix] - 1)
            else:
                output += dict[prefix] * dict[suffix]
        return output
        
                    
                

class Solution:
    def numOfPairs(self, nums: List[str], target: str) -> int:
        freq = Counter(nums)
        output = 0
        
        # freq
        # Counter({'77': 2, '777': 1, '7': 1}) 
        # freq.items()
        # dict_items([('777', 1), ('7', 1), ('77', 2)])
        
        for k, v in freq.items():
            # prefix target[0:len(k)]
            if target.startswith(k):
                suffix = target[len(k):]
                # print(target[0:len(k)], suffix, v, freq[suffix])
                output += v * freq[suffix]
                # prefix == suffix
                if k == suffix: 
                    output -= freq[suffix]
        return output
                    
                

File: Brute-Force/2023. Number of Pairs of Strings With Concatenation Equal to Target/numOfPairs.js
/**
 * @param {string[]} nums
 * @param {string} target
 * @return {number}
 */
// brute force
 var numOfPairs = function(nums, target) {
    let output = 0;
    
    for (let i = 0; i < nums.length; i++) {
        for (let j = 0; j < nums.length; j++) {
            if (nums[i] + nums[j] === target && j !== i) {
                output += 1
            }  
        }
    }
    
    return output;
};

// Map
/**
 * @param {string[]} nums
 * @param {string} target
 * @return {number}
 */
 var numOfPairs = function(nums, target) {
    let output = 0;
    
    let map = new Map();
    for (let num of nums) {
        map.set(num, (map.get(num) || 0) + 1);
    }
    
    for (let i = 0; i < target.length; i++) {
        // split target
        let prefix = target.substring(0, i);
        let suffix = target.substring(i, target.length);
        
        // calculate output
        if (map.has(prefix) && map.has(suffix)) {
            if (prefix === suffix) {
                output += map.get(prefix) * (map.get(suffix) - 1);
            } else {
                output += map.get(prefix) * map.get(suffix);
            }
        }
        
    }
    
    return output;
};